Dash / Tunnel Docs (#81)
moj-analytical-services/platform_user_guidance: 16-Access-Local-Ports-in-Jupyter.Rmd

```{r global_options, include=FALSE}
{knitr::opts_chunk$set(eval = FALSE)}
```
# Access Local Ports in Jupyter
If you are developing a web application using Jupyter you will probably want to

> preview that webapp before you deploy it. The platform provides a special set of

Reviewer: ```suggestion
preview that web app before you deploy it. The Analytical Platform provides a special set of
```

> endpoints under `/_tunnnel_/<portNumber>/` that route to exposed HTTP ports on

Reviewer: ```suggestion
endpoints under `/_tunnel_/<portNumber>/` that route to exposed HTTP ports on
```

the localhost. There are a few things you need to keep in mind:

> * `<portNumber>` can only be one from a limited rage. By default you can only route

Reviewer: ```suggestion
* `<portNumber>` can only be one from a limited range. By default, you can only route
```

>   to ports 8050 and 4040-4050 inclusive. You should make your webapp listen on

Reviewer: ```suggestion
  to ports 8050 and 4040--4050 inclusive. You should make your web app listen on
```

  one of those.
* The `/_tunnel_/<portNumber>/` endpoint will only tunnel to services bound to a

>   non public IP address. By default many web frameworks bind to host `127.0.0.1`

Reviewer: ```suggestion
  non--public IP address. By default, many web frameworks bind to host `127.0.0.1`.
```

>   you will need to change this to `0.0.0.0` or the tunnel won't work. This is to

Reviewer: ```suggestion
  You will need to change this to `0.0.0.0` or the tunnel won't work. This is to
```

  prevent inadvertently exposing webapps to the tunnel.

## Running Plotly [Dash] apps

### Install dependencies

> In the terminal install the dash dependencies:

Reviewer: ```suggestion
In the terminal, install the Dash dependencies:
```

```bash
pip install --user dash==0.39.0  # The core dash backend
pip install --user dash-daq==0.1.0  # DAQ components (newly open-sourced!)
```

The demo code in this document is known to work with these versions but you can
install any version you like for your own code.

![](images/dash/dash_install_deps.gif)

> If you are planning on turning this into a project you'll need to manage your

Reviewer: ```suggestion
If you are planning on turning this into a project, you'll need to manage your
```

> dependencies by adding these to either `requirements.txt` & `pip` or

Reviewer: ```suggestion
dependencies by adding these to either `requirements.txt` and `pip` or
```

> `environment.yaml` & `conda`.

Reviewer: ```suggestion
`environment.yaml` and `conda`.
```

### Example code

> Save the dash hello world to a new python file called `app.py`:

Reviewer: ```suggestion
Save the Dash hello world app to a new python file called `app.py`:
```

app.scripts.config.serve_locally = False
dcc._js_dist[0]['external_url'] = 'https://cdn.plot.ly/plotly-basic-latest.min.js'

app.layout = html.Div([
    html.H1('Stock Tickers'),
    dcc.Dropdown(
        id='my-dropdown',
        options=[
            {'label': 'Tesla', 'value': 'TSLA'},
            {'label': 'Apple', 'value': 'AAPL'},
            {'label': 'Coke', 'value': 'COKE'}
        ],
        value='TSLA'
    ),
    dcc.Graph(id='my-graph')
], className="container")

@app.callback(Output('my-graph', 'figure'),
              [Input('my-dropdown', 'value')])

def update_graph(selected_dropdown_value):
    dff = df[df['Stock'] == selected_dropdown_value]
    return {
        'data': [{
            'x': dff.Date,
            'y': dff.Close,
            'line': {
                'width': 3,
                'shape': 'spline'
            }
        }],
        'layout': {
            'margin': {
                'l': 30,
                'r': 20,
                'b': 30,
                't': 20
            }
        }
    }

if __name__ == '__main__':
    app.run_server(host='0.0.0.0')

```

There are two important things to note here, where this code differs from the

> plotly dash example:

Reviewer: ```suggestion
Plotly Dash example:
```

> The `Dash` class must be instanciated with a `url_base_pathname`. This should

Reviewer 2: ```suggestion
The `Dash` class must be instantiated with a `url_base_pathname`. This should
```

Author: good spot :pray:

> The `Dash` class must be instantiated with a `url_base_pathname`. This should

Reviewer: ```suggestion
1. The `Dash` class must be instantiated with a `url_base_pathname`. This should
```

always be `/_tunnel_/8050/` e.g.
`app = dash.Dash('app', server=server, url_base_pathname='/_tunnel_/8050/')`.

> When you run the server it must be bound to a `0.0.0.0` e.g. `app.run_server(host='0.0.0.0')`

Reviewer: ```suggestion
2. When you run the server, it must be bound to `0.0.0.0`, e.g., `app.run_server(host='0.0.0.0')`.
```

![](images/dash/save_example.gif)

### Run server from the terminal

```sh
yovyan@jupyter-lab-r4vi-jupyter-7cb4bb58cf-6hngf:~$ python3 app.py
 * Serving Flask app "app" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://0.0.0.0:8050/ (Press CTRL+C to quit)
```

> ### Access via `_tunnel_` url

Reviewer: ```suggestion
### Access via `_tunnel_` URL
```

> Copy your `jupyter` URL and append `/_tunnel_/8050/` to access your running dash

Reviewer: ```suggestion
Copy your `jupyter` URL and append `/_tunnel_/8050/` to access your running Dash
```

> app, if you're `jupyter` url is: `https://r4vi-jupyter-lab.tools.alpha.mojanalytics.xyz/lab?`

Reviewer: ```suggestion
app. If you're `jupyter` URL is `https://r4vi-jupyter-lab.tools.alpha.mojanalytics.xyz/lab?`,
```
